internal/build/extend_build_test.go

```diff
@@ -0,0 +1,160 @@
+		providedPublish               bool
+		providedBuilderImage          = "some-registry.com/some-namespace/some-builder-name"
+		extendedBuilderImage          = "some-registry.com/some-namespace/some-builder-name-extended"
+		configureDefaultTestLifecycle func(opts *build.LifecycleOptions)
+		lifecycleOps                  []func(*build.LifecycleOptions)
+	)
+
+	it.Before(func() {
+		var err error
+		mockController = gomock.NewController(t)
+		mockDockerClient = mockdocker.NewMockDockerClient(mockController)
+		h.AssertNil(t, err)
+
+		configureDefaultTestLifecycle = func(opts *build.LifecycleOptions) {
+			opts.BuilderImage = providedBuilderImage
+			opts.ClearCache = providedClearCache
+			opts.Publish = providedPublish
+		}
+
+		lifecycleOps = []func(*build.LifecycleOptions){configureDefaultTestLifecycle}
+	})
+
+	when("Extend Build Image By Docker", func() {
+		it("should extend build image using 1 extension", func() {
+			// set tmp directory
+			tmpDir = "./testdata/fake-tmp/build-extension/single"
+			lifecycle = getTestLifecycleExec(t, true, tmpDir, mockDockerClient, lifecycleOps...)
+			expectedBuilder := lifecycle.Builder()
+			expectedBuildContext := archive.ReadDirAsTar(filepath.Dir(filepath.Join(tmpDir, "fake-tmp", "build-extension", "single", "build", "samples_test", "Dockerfile")), "/", 0, 0, -1, true, false, func(file string) bool { return true })
+			// Set up expected Build Args
+			UID := strconv.Itoa(expectedBuilder.UID())
+			GID := strconv.Itoa(expectedBuilder.GID())
+			expectedbuildArguments := map[string]*string{}
+			expectedbuildArguments["base_image"] = &providedBuilderImage
+			expectedbuildArguments[argUserID] = &UID
+			expectedbuildArguments[argGroupID] = &GID
+			expectedBuildOptions := types.ImageBuildOptions{
+				Context:    expectedBuildContext,
+				Dockerfile: "Dockerfile",
+				Tags:       []string{extendedBuilderImage},
+				Remove:     true,
+				BuildArgs:  expectedbuildArguments,
+			}
+			mockResponse := types.ImageBuildResponse{
+				Body:   io.NopCloser(strings.NewReader("mock-build-response-body")),
+				OSType: "linux",
+			}
+			mockDockerClient.EXPECT().ImageBuild(gomock.Any(), gomock.Any(), gomock.Any()).Do(func(_ context.Context, buildContext io.ReadCloser, buildOptions types.ImageBuildOptions) {
+				compBuildOptions(t, expectedBuildOptions, buildOptions)
+			}).Return(mockResponse, nil).Times(1)
+			err := lifecycle.ExtendBuildByDaemon(context.Background())
+			h.AssertNil(t, err)
+		})
+
+		it("should extend build image using multiple extension", func() {
+			// set tmp directory
+			tmpDir = "./testdata/fake-tmp/build-extension/multi"
+			lifecycle = getTestLifecycleExec(t, true, tmpDir, mockDockerClient, lifecycleOps...)
+			mockResponse := types.ImageBuildResponse{
+				Body:   io.NopCloser(strings.NewReader("mock-build-response-body")),
+				OSType: "linux",
+			}
+			mockDockerClient.EXPECT().ImageBuild(gomock.Any(), gomock.Any(), gomock.Any()).Return(mockResponse, nil).Times(2)
```

Comment:
Ideally here we'd also want to check that the options provided to the second Dockerfile are correct. `Dockerfile` and `BuildArgs` seem to be the most important to test.

Reply:
@itsdarshankumar please take a look at this suggestion, I think is very valid to make test useful

Reply:
yes eyes on it :)